lua/lvim/core/luasnip/init.lua

```diff
@@ -0,0 +1,59 @@
+local M = {}
+function M.config()
+  lvim.builtin.luasnip = {
+    sources = {
+      friendly_snippets = true,
+      lunarvim = true,
+    },
+    config = {
+      history = false,
+      updateevents = "TextChanged,TextChangedI",
+      ext_opts = {
+        -- Will be populated within config function
+      },
+    },
+  }
+end
+
+function M.setup()
+  local utils = require "lvim.utils"
+  local paths = {}
+  if lvim.builtin.luasnip.sources.friendly_snippets then
+    paths[#paths + 1] = utils.join_paths(get_runtime_dir(), "site", "pack", "packer", "start", "friendly-snippets")
+  end
+  local user_snippets = utils.join_paths(get_config_dir(), "snippets")
+  if utils.is_directory(user_snippets) then
+    paths[#paths + 1] = user_snippets
+  end
+  -- When no paths are provided, luasnip will search in the runtimepath
+  require("luasnip.loaders.from_lua").lazy_load()
+  require("luasnip.loaders.from_vscode").lazy_load {
+    paths = paths,
+  }
+  require("luasnip.loaders.from_snipmate").lazy_load()
+
+  local luasnip = require "luasnip"
+  local types = require "luasnip.util.types"
+
+  local ext_opts = {
+    -- Show virtual text to signal when you are inside an sippets
+    [types.insertNode] = {
+      active = {
+        virt_text = { { "<-- snip insert", "BufferInactiveIndex" } },
+      },
+    },
+    -- Helps to notice when you are within a choice node
+    [types.choiceNode] = {
+      active = {
+        virt_text = { { "<-- choice", "BufferInactiveIndex" } },
+      },
+    },
+  }
+  -- Add lunarvim options giving preference to user ones
+  lvim.builtin.luasnip.config.ext_opts = vim.tbl_deep_extend("keep", lvim.builtin.luasnip.config.ext_opts, ext_opts)
+  luasnip.config.set_config(lvim.builtin.luasnip.config)
+  if lvim.builtin.luasnip.sources.lunarvim then
+    luasnip.add_snippets("lua", require "lvim.core.luasnip.snippets")
+  end
```

Comment:
- I'm not sure this will be clear enough, we probably need to make a distinction between `lua` and `json` snippets.
- is it possible to lock a snippet to a folder/file? for lvim specific

Reply:
> * I'm not sure this will be clear enough, we probably need to make a distinction between `lua` and `json` snippets.

What do you mean by it not being clear enough? To who it needs to be clear? This is not user facing. I think any dev working with this part of the app will understand this is a call to one of the LuaSnip APIs to add the snippets referenced in the require.

> * is it possible to lock a snippet to a folder/file? for lvim specific

I'm not sure if I understand correctly what you mean. But if you mean to only add lunar-vim snippets when you enter/open lunar-vim folders I think we can add an autocomand on buf-enter that checks the path and adds the snippets if it matches one of the lvim runtime paths or config paths or something like that, but it may be tricky for users that do not use the regular folder path (just like me). I thought about that, and I came to the conclusion that it is better to just let the user disable them if they find them not useful.

Reply:
> What do you mean by it not being clear enough?

judging by how you're using it, `luasnip.sources.lunarvim` is responsible for loading `lua` snippets as well as `json`-style ones. I'm not sure what an easy way would be to differentiate the two cases.

> but it may be tricky for users that do not use the regular folder path

That should always be covered with `$LUNARVIM_CONFIG_DIR`, I don't see how it wouldn't be. Although, I was hoping for a native option in luasnip, as anything else would be more effort than warranted.

Reply:
> judging by how you're using it, `luasnip.sources.lunarvim` is responsible for loading `lua` snippets as well as `json`-style ones. I'm not sure what an easy way would be to differentiate the two cases.
> 
> >

Sorry, I have been reading the luasnips docs recently and I assume a familiarity that probably not everybody has. Luasnip has a bunch of mechanisms to load snippets, and this is just one of them.
There are the snippet loaders, which are the most common and convenient way to load snippets into luansip.
You can also parametrize those loaders with additional paths for luasnip to load. This is what we do to load the snippets from the friendly snippets folder. If you don't provide any path, the loader will automatically search in a bunch of places within the runtimepath.
Lastly, [there is the programmatic API](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#api-reference) (which is internally used by the loader) that allows you to pass one or two tables containing snippets. This last method is what I am using here. The reason is because it is more convenient this way. We don't have to mess with concatenating/generating paths that may vary drastically depending on the user environment. Since they are supposed to be built into lunarvim we can just require them and add them in case the user wants to enable them via config.
This has nothing to do with the users adding their own snippets, they can always call the different methods (the API, the different loaders) with the parameters they want to additionally load their snippets from whenever they want. I don't see a reason to abstract that away.

Hope this makes sense.

Reply:
By the way, I recently added docs to LunarVim about how to load snippets from Lua. That is still valid: https://www.lunarvim.org/configuration/08-custom-snippets.html#lua-version